pkg/kafka/policy.go

```diff
@@ -0,0 +1,219 @@
+		}
+	}
+
+	return false
+}
+
+func matchProduceReq(req *proto.ProduceReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !produceTopicContained(rule.Topic, req.Topics) {
+		log.Debugf("denied (topic)")
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	log.Debugf("allowed produce")
+	return true
+}
+
+func fetchTopicContained(neededTopic string, topics []proto.FetchReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchFetchReq(req *proto.FetchReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !fetchTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func offsetTopicContained(neededTopic string, topics []proto.OffsetReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetReq(req *proto.OffsetReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !offsetTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func topicContained(neededTopic string, topics []string) bool {
+	for _, topic := range topics {
+		if topic == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchMetadataReq(req *proto.MetadataReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !topicContained(rule.Topic, req.Topics) {
+		log.Debugf("Denied (topic)")
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	log.Debugf("allowed metadata")
+	return true
+}
+
+func offsetCommitTopicContained(neededTopic string, topics []proto.OffsetCommitReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetCommitReq(req *proto.OffsetCommitReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !offsetCommitTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func offsetFetchTopicContained(neededTopic string, topics []proto.OffsetFetchReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetFetchReq(req *proto.OffsetFetchReq, rule api.PortRuleKafka) bool {
+	if rule.Topic != "" && !offsetFetchTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
+	log.Debugf("Matching rule %q", rule)
+	log.Debugf("Required API key: %d", rule.GetAPIKey())
+	if rule.GetAPIKey() != -1 {
```

Comment:
This might be simpler to read as a switch.
```golang
switch apiKey := rule.GetAPIKey(); apiKey {
case -1:
	// do nothing
case req.kind:
	log.Debugf("API key matches")
default:
	log.Debugf("denied (apikey)")
	return false
}
```

Reply:
Agreed, fixed